File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/SprayForecastWindows.kt
package com.rork.vinetrack.data

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/** Vineyard-local four-hour forecast in metric units. Null means unknown, never zero. */
data class SprayForecastPeriod(
    val date: String,
    val startHour: Int,
    val startTimeLocal: String = "%02d:00".format(Locale.US, startHour),
    val endTimeLocal: String = "%02d:00".format(Locale.US, startHour + 4),
    val tempMinC: Double? = null,
    val tempMaxC: Double? = null,
    val windMaxKmh: Double? = null,
    val humidityMinPct: Double? = null,
    val humidityMaxPct: Double? = null,
    val rainMm: Double? = null,
    val sampleCount: Int = 0,
) {
    fun fillingNulls(other: SprayForecastPeriod): SprayForecastPeriod =
        if (date != other.date || startHour != other.startHour) this else copy(
            tempMinC = tempMinC ?: other.tempMinC,
            tempMaxC = tempMaxC ?: other.tempMaxC,
            windMaxKmh = windMaxKmh ?: other.windMaxKmh,
            humidityMinPct = humidityMinPct ?: other.humidityMinPct,
            humidityMaxPct = humidityMaxPct ?: other.humidityMaxPct,
            rainMm = rainMm ?: other.rainMm,
        )
}

data class SprayForecastWindow(val start: Long, val end: Long, val kind: Kind) {
    enum class Kind { OPTIMAL, HIGH_HUMIDITY }

    fun label(timezone: TimeZone, now: Long): String {
        val clock = SimpleDateFormat("h:mm a", Locale.ENGLISH).apply { timeZone = timezone }
        val weekday = SimpleDateFormat("EEE", Locale.ENGLISH).apply { timeZone = timezone }
        val calendar = Calendar.getInstance(timezone)
        fun dayKey(time: Long): String = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply { timeZone = timezone }.format(Date(time))
        val today = dayKey(now)
        calendar.timeInMillis = now
        calendar.add(Calendar.DAY_OF_YEAR, 1)
        val heading = when (dayKey(start)) {
            today -> "Today"
            dayKey(calendar.timeInMillis) -> "Tomorrow"
            else -> weekday.format(Date(start))
        }
        val range = if (dayKey(start) == dayKey(end - 1)) {
            "${clock.format(Date(start))} – ${clock.format(Date(end))}"
        } else {
            "${weekday.format(Date(start))} ${clock.format(Date(start))} – ${weekday.format(Date(end))} ${clock.format(Date(end))}"
        }
        return "$heading\n$range ${if (kind == Kind.OPTIMAL) "Optimal" else "High humidity"}"
    }
}

/** Provider-independent Portal-default qualification and non-overlapping display bands. */
object SprayForecastWindows {
    fun qualifiesOptimal(p: SprayForecastPeriod): Boolean {
        val low = p.tempMinC ?: return false
        val high = p.tempMaxC ?: return false
        val wind = p.windMaxKmh ?: return false
        val rain = p.rainMm ?: return false
        return low > 10 && high < 35 && wind < 15 && rain <= 0.1
    }

    fun qualifiesHighHumidity(p: SprayForecastPeriod): Boolean =
        qualifiesOptimal(p) && (p.humidityMinPct?.let { it >= 90 } == true)

    fun supplement(primary: List<SprayForecastPeriod>, secondary: List<SprayForecastPeriod>): List<SprayForecastPeriod> {
        val byKey = secondary.associateBy { it.date to it.startHour }
        val existing = primary.map { it.date to it.startHour }.toSet()
        return primary.map { p -> byKey[p.date to p.startHour]?.let(p::fillingNulls) ?: p } +
            secondary.filter { (it.date to it.startHour) !in existing }
    }

    fun windows(periods: List<SprayForecastPeriod>, timezone: TimeZone, now: Long = System.currentTimeMillis()): List<SprayForecastWindow> {
        val calendar = Calendar.getInstance(timezone).apply {
            timeInMillis = now
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        val first = calendar.timeInMillis
        calendar.add(Calendar.DAY_OF_YEAR, 5)
        val last = calendar.timeInMillis
        val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply { timeZone = timezone; isLenient = false }
        val slots = periods.mapNotNull { p ->
            if (p.startHour !in 0..23 || p.startHour % 4 != 0 || !qualifiesOptimal(p)) return@mapNotNull null
            val date = parser.parse(p.date) ?: return@mapNotNull null
            if (date.time < first || date.time >= last) return@mapNotNull null
            calendar.time = date
            calendar.set(Calendar.HOUR_OF_DAY, p.startHour)
            val start = calendar.timeInMillis
            calendar.time = date
            if (p.startHour == 20) calendar.add(Calendar.DAY_OF_YEAR, 1)
            else calendar.set(Calendar.HOUR_OF_DAY, p.startHour + 4)
            val end = calendar.timeInMillis
            if (end <= now) null else SprayForecastWindow(start, end,
                if (qualifiesHighHumidity(p)) SprayForecastWindow.Kind.HIGH_HUMIDITY else SprayForecastWindow.Kind.OPTIMAL)
        }.sortedBy { it.start }
        val result = mutableListOf<SprayForecastWindow>()
        for (slot in slots) {
            val previous = result.lastOrNull()
            if (previous != null && previous.end == slot.start && previous.kind == slot.kind) {
                result[result.lastIndex] = previous.copy(end = slot.end)
            } else result.add(slot)
        }
        return result
    }
}

File: ios/VineTrack/App/SprayForecastWindows.swift
import Foundation

/// Vineyard-local four-hour forecast, always in metric units. Nil is unknown, not zero.
nonisolated struct SprayForecastPeriod: Sendable, Hashable {
    let date: String
    let startHour: Int
    let startTimeLocal: String
    let endTimeLocal: String
    let tempMinC: Double?
    let tempMaxC: Double?
    let windMaxKmh: Double?
    let humidityMinPct: Double?
    let humidityMaxPct: Double?
    let rainMm: Double?
    let sampleCount: Int

    func fillingNulls(from other: Self) -> Self {
        guard date == other.date, startHour == other.startHour else { return self }
        return Self(date: date, startHour: startHour, startTimeLocal: startTimeLocal,
                    endTimeLocal: endTimeLocal, tempMinC: tempMinC ?? other.tempMinC,
                    tempMaxC: tempMaxC ?? other.tempMaxC, windMaxKmh: windMaxKmh ?? other.windMaxKmh,
                    humidityMinPct: humidityMinPct ?? other.humidityMinPct,
                    humidityMaxPct: humidityMaxPct ?? other.humidityMaxPct,
                    rainMm: rainMm ?? other.rainMm, sampleCount: sampleCount)
    }
}

nonisolated struct SprayForecastWindow: Sendable, Equatable {
    enum Kind: Sendable { case optimal, highHumidity }
    let start: Date
    let end: Date
    let kind: Kind

    func label(in timezone: TimeZone, now: Date) -> String {
        let clock = DateFormatter()
        clock.locale = Locale(identifier: "en_AU")
        clock.timeZone = timezone
        clock.dateFormat = "h:mm a"
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timezone
        let day = DateFormatter()
        day.locale = Locale(identifier: "en_AU")
        day.timeZone = timezone
        day.dateFormat = "EEE"
        let today = calendar.isDate(start, inSameDayAs: now)
        let tomorrow = calendar.isDate(start, inSameDayAs: calendar.date(byAdding: .day, value: 1, to: now) ?? now)
        let heading = today ? "Today" : tomorrow ? "Tomorrow" : day.string(from: start)
        let range = calendar.isDate(start, inSameDayAs: end.addingTimeInterval(-1))
            ? "\(clock.string(from: start)) – \(clock.string(from: end))"
            : "\(day.string(from: start)) \(clock.string(from: start)) – \(day.string(from: end)) \(clock.string(from: end))"
        return "\(heading)\n\(range) \(kind == .optimal ? "Optimal" : "High humidity")"
    }
}

/// Provider-independent Portal-default qualification and non-overlapping display bands.
nonisolated enum SprayForecastWindows {
    static func qualifiesOptimal(_ p: SprayForecastPeriod) -> Bool {
        guard let low = p.tempMinC, let high = p.tempMaxC,
              let wind = p.windMaxKmh, let rain = p.rainMm else { return false }
        return low > 10 && high < 35 && wind < 15 && rain <= 0.1
    }

    static func qualifiesHighHumidity(_ p: SprayForecastPeriod) -> Bool {
        qualifiesOptimal(p) && (p.humidityMinPct.map { $0 >= 90 } ?? false)
    }

    static func supplement(_ primary: [SprayForecastPeriod], with secondary: [SprayForecastPeriod]) -> [SprayForecastPeriod] {
        let values = Dictionary(secondary.map { ("\($0.date):\($0.startHour)", $0) }, uniquingKeysWith: { first, _ in first })
        let existing = Set(primary.map { "\($0.date):\($0.startHour)" })
        return primary.map { p in
            guard let other = values["\(p.date):\(p.startHour)"] else { return p }
            return p.fillingNulls(from: other)
        } + secondary.filter { !existing.contains("\($0.date):\($0.startHour)") }
    }

    static func windows(_ periods: [SprayForecastPeriod], timezone: TimeZone, now: Date = Date()) -> [SprayForecastWindow] {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timezone
        let firstDay = calendar.startOfDay(for: now)
        guard let lastDay = calendar.date(byAdding: .day, value: 5, to: firstDay) else { return [] }
        let parser = DateFormatter()
        parser.locale = Locale(identifier: "en_US_POSIX")
        parser.timeZone = timezone
        parser.dateFormat = "yyyy-MM-dd"
        let slots: [(Date, Date, SprayForecastWindow.Kind)] = periods.compactMap { p in
            guard (0..<24).contains(p.startHour), p.startHour % 4 == 0,
                  let date = parser.date(from: p.date), date >= firstDay, date < lastDay,
                  let start = calendar.date(bySettingHour: p.startHour, minute: 0, second: 0, of: date),
                  let end = p.startHour == 20
                    ? calendar.date(byAdding: .day, value: 1, to: date)
                    : calendar.date(bySettingHour: p.startHour + 4, minute: 0, second: 0, of: date),
                  end > now,
                  qualifiesOptimal(p) else { return nil }
            return (start, end, qualifiesHighHumidity(p) ? .highHumidity : .optimal)
        }.sorted { $0.0 < $1.0 }
        var result: [SprayForecastWindow] = []
        for (start, end, kind) in slots {
            if let last = result.last, last.end == start, last.kind == kind {
                result[result.count - 1] = SprayForecastWindow(start: last.start, end: end, kind: kind)
            } else {
                result.append(SprayForecastWindow(start: start, end: end, kind: kind))
            }
        }
        return result
    }
}
